### ISSUE TITLE
`editor::SwitchSourceHeader` LSP action doesn't work in SSH remote

### ISSUE THREAD

**Issue by shahamran** [zed-industries/zed#20583]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Add this to `keymap.json`:
   ```json
   {
     "use_layout_keys": true,
     "context": "Editor",
     "bindings": {
       "alt-o": "editor::SwitchSourceHeader"
     }
   }
   ```
2. Open a ssh remote with a C/C++ project.
3. Press `alt-o` in a header file.
4. `<nothing happens>`

The same _works_ in a local (non-remote) C project.

Also, the LSP logs pane stays empty in remote, not sure if it's related.

### Environment

Zed: v0.160.7 (Zed)
OS: macOS 14.7.0
Memory: 32 GiB
Architecture: aarch64

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

```

```
</details>

**Comment by notpeter** [comment]
Do you have clangd installed on the remote?
Do other language server functions work (auto-complete, etc)?
Is there anything in your Zed Logs when this happens?

**Comment by shahamran** [comment]
Nothing relevant in zed logs, nothing at all in lsp logs, other LSP functionality works fine (and same request works in VSCode with the same clangd executable).

**Comment by notpeter** [comment]
I know there were some recent improvements with remoting and LSP Logs.  Later today we'll be releasing a new Zed Preview (v0.162.x) could you try that and report back if things have improved?

**Comment by shahamran** [comment]
@notpeter the situation is exactly the same in 162.0. Empty LSP logs, nothing related in Zed.log.

**Comment by Poldraunic** [comment]
This **does not** for me in the local repo. Should I create a new issue for that?

**Comment by notpeter** [comment]
@Poldraunic / @shahamran Can you provide some steps for reproduction, ideally with an open source project or a small set of minimal files where it works in VSCode, but not in Zed?

I wonder whether there was a regression since:
- https://github.com/zed-industries/zed/issues/12801

Or if there is there's something distinct to your setup which Zed needs to support.

**Comment by Poldraunic** [comment]
Oh, and as you can see from the file names, I am on Windows. I know it is not officially supported so this might be why it doesn't work for me.

**Comment by shahamran** [comment]
@notpeter

> Does it work for you locally with Zed?

Yes, I mentioned it in the original post.

> Can you provide some steps for reproduction, ideally with an open source project or a small set of minimal files where it works in VSCode, but not in Zed?

Sure. I tested with https://github.com/munificent/craftinginterpreters.

1. Open `c/compiler.h`.
2. Trigger `editor::SwitchSourceHeader`.
3. In SSH remote: nothing happens (and nothing in the logs).
   In local Zed: a new buffer opens with `c/compiler.c`.

> I wonder whether there was a regression since: ...

I wouldn't know, but my guess is that it was never fully tested in SSH remote context.

**Comment by notpeter** [comment]
@Poldraunic Thanks for the simple repo. Can you open a distinct issue for `editor::SwitchSourceHeader` being broken on windows.

@shahamran Apologies for asking for info you'd already provided. My bad.

I was able to reproduce, `Editor::SwitchSourceHeader` works locally but the action is not available when connected via SSH.

Steps I used to reproduce:

```
mkdir -p ~/source
cd ~/source
git clone https://github.com/Poldraunic/zed-20583
zed ssh://localhost/~/source/zed-20583
```

Open `MyClass.cpp`
Clangd works perfectly
But `cmd-shift-p` and no `editor: switch source header` action available.

Thanks all for helping getting this repro nailed down.

**Comment by Poldraunic** [comment]
@notpeter absolutely! I'll take a crack at fixing it myself first and if I am unable, I will make an issue.

**Comment by SomeoneToIgnore** [comment]
The culprit is this line: https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/editor/src/lsp_ext.rs#L40

We do not really sync the metadata (including LSP adapters) about language servers to the remote clients, hence this call will always return an empty list and we will never find any language server that we need.

As a hacky workaround, we may check https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/project/src/lsp_store.rs#L716 instead, as there should be server names, but not sure how would this work in reality.

**Comment by notpeter** [comment]
Thanks for fixing the windows issue in #20856 @Poldraunic.

**Comment by shahamran** [comment]
Still happens in latest version of Zed.

> The culprit is this line:
> 
> [zed/crates/editor/src/lsp_ext.rs](https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/editor/src/lsp_ext.rs#L40)
> 
> Line 40 in [81ca890](/zed-industries/zed/commit/81ca890a2a65881b4526faeb4526b9c14c1228d5)
> 
>  .language_servers_for_buffer(buffer.read(cx), cx) 
> We do not really sync the metadata (including LSP adapters) about language servers to the remote clients, hence this call will always return an empty list and we will never find any language server that we need.
> 
> As a hacky workaround, we may check
> 
> [zed/crates/project/src/lsp_store.rs](https://github.com/zed-industries/zed/blob/81ca890a2a65881b4526faeb4526b9c14c1228d5/crates/project/src/lsp_store.rs#L716)
> 
> Line 716 in [81ca890](/zed-industries/zed/commit/81ca890a2a65881b4526faeb4526b9c14c1228d5)
> 
>  pub language_server_statuses: BTreeMap<LanguageServerId, LanguageServerStatus>, 
> instead, as there should be server names, but not sure how would this work in reality.

I wanted to work on this but I'm afraid I don't have enough background knowledge about LSPs and Zed's architecture.
@SomeoneToIgnore I would appreciate guidance if you think this is something suitable for a Zed-beginner.
Thanks!

**Comment by tuxmark5** [comment]
If this is technically difficult to solve, I'd appreciate a workaround of unconditionally enabling certain actions with no regard of current file/project type. For me personally fixing #27133 is vital to be able to use Zed in our proc_macro heavy project.

**Comment by SomeoneToIgnore** [comment]
This looks like a moderate task, potentially simple but may be simple, not sure as requires some research first.

The main idea is: 

* we need to adjust `project.language_server_id_for_name` call somehow in

https://github.com/zed-industries/zed/blob/6715c8582f3915c352e1a983c1c150c67309c204/crates/editor/src/lsp_ext.rs#L32-L49

* there's a set of language server metadata (name, id, etc.) associated with each buffer, but it exists only where project's FS is.
 
* there's no dedicated "sync langserver metadata" activity yet, so the remote clients do not know much except maybe `language_server_statuses` and other things that already exist for some other purpose.
 
* a buffer may have N languages inside it (e.g. foo.ts script with Typescript as a "main" language and comments having markdown with code snippets of html language) and M language servers associated with each of the language (*.ts may require typescript-language-server for general TypeScript support and another one for Angular support, or Vue and/or eslint, or ...)
 
* when an action is triggered, the client knows a `buffer_id` and `language` relevant for the bullet above, and needs to get an associated language server with a particular name (rust-analyzer, clangd). 

* both Zed with FS and clients share the same `buffer_id` and `language` knowledge, but the client knows almost nothing about the `server_id` and other server-related metadata

First, we need to understand whether remote client's `LspStore` has a way to connect `buffer_id` with a `server_id` and `LanguageServerName` data.
If there is one, we're changing `fn language_server_id_for_name` logic in the `Project` to `if self.is_local() { ..old code } else { ..new code }` and that's it for the fix.
Would be great to add a test similar to https://github.com/zed-industries/zed/blob/6715c8582f3915c352e1a983c1c150c67309c204/crates/collab/src/tests/editor_tests.rs#L281 but about `rust-analyzer`.

If not, we need to think more, but, if results are elaborately stated, working up to here is already a good contribution compared to any of the comments in this issue so far.

----------------

If not, we need to invent a way to send that `server_id` from the FS part to the remote client part, there's https://github.com/zed-industries/zed/blob/main/crates/proto/proto/zed.proto file and one can check its file's history and see what needs to be changed for adding another method.
Or, you can check for e.g. `GetCodeLens` usage and see all the places that need to be changed.

Seems that we can replicate `language_server_id_for_name` in zed.proto and that is enough?
I would remove `id` mention from the proto declaration, just in case we would need to send more metadata later, not just the ID.
